main.cc

```diff
@@ -146,6 +146,62 @@ read_config(bpo::variables_map& opts, db::config& cfg) {
         return make_exception_future<>(ep);
     });
 }
+
+class sighup_handler {
+    bpo::variables_map& _opts;
+    db::config& _cfg;
+    condition_variable _cond;
+    bool _pending = false; // if asked to reread while already reading
+    bool _stopping = false;
+    bool _stopped = false;
+    future<> _done = do_work();
+public:
+    sighup_handler(bpo::variables_map& opts, db::config& cfg) : _opts(opts), _cfg(cfg) {
+        startlog.info("installing SIGHUP handler");
+        engine().handle_signal(SIGHUP, [this] { reread_config(); });
+    }
+private:
+    void reread_config() {
+        if (_stopping) {
+            return;
+        }
+        _pending = true;
+        _cond.broadcast();
+    }
+    future<> do_work() {
+        return repeat([this] {
+            return _cond.wait([this] { return _pending || _stopping; }).then([this] {
+                return async([this] {
+                    if (_stopping) {
+                        _stopped = true;
+                        return stop_iteration::yes;
+                    } else if (_pending) {
+                        _pending = false;
+                        try {
+                            startlog.info("re-reading configuration file");
+                            read_config(_opts, _cfg).get();
+                            _cfg.broadcast_to_all_shards().get();
+                            startlog.info("completed re-reading configuration file");
+                        } catch (...) {
+                            startlog.error("failed to re-read configuration file: {}", std::current_exception());
+                        }
+                    }
+                    return stop_iteration::no;
+                });
+            });
+        });
+    }
+public:
+    future<> stop() {
```

Comment:
A blurb would be helpful here.  When should this be called, and what is it intended to return?

Reply:
added